### PR TITLE
Add feature window.prompt

### PR DESCRIPTION
**Pull request by AlexRobinson-**
Fixes https://github.com/brave/browser-laptop/issues/94

## What it looks like
window-prompt-highlight-text

## Submitter Checklist:
- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request [as needed](https://github.com/brave/browser-laptop/wiki/Pull-request-process).
- [ ] Request a security/privacy review [as needed](https://github.com/brave/handbook/blob/master/development/security.md#how-to-request-a-security-review). (Ask a Brave employee to help if you cannot access this document.)

## Test Plan:
See https://github.com/brave/browser-laptop/issues/94

## Automated test plan
`npm run unittest -- --grep="onWindowPrompt"`
`npm run unittest -- --grep="MessageBox component unit tests"`

## Reviewer Checklist:

- [ ] Request a security/privacy review [as needed](https://github.com/brave/handbook/blob/master/development/security.md#how-to-request-a-security-review) if one was not already requested.

Tests
- [x] Adequate test coverage exists to prevent regressions
- [x] Tests should be independent and work correctly when run individually or as a suite [ref](https://github.com/brave/browser-laptop/wiki/Code-Guidelines#test-dependencies)
- [x] New files have MPL2 license header

### CODE REVIEW

**Comment by ryanml** [comment]
@AlexRobinson- what a small world... I started on this implementation last night and our approaches are very close! Really solid job! I'm excited to see this get implemented :D

**Comment by AlexRobinson-** [comment]
Oh no, I'm sorry for stealing this from you @ryanml ! Are there any parts of your implementation you prefer over mine you would like to see go into the PR?

**Comment by ryanml** [comment]
Oh no not at all @AlexRobinson- ! I didn't call it any way, just was a coincidence :) And honestly aside from naming we did just about the same thing

**Comment by bradleyrichter** [comment]
@AlexRobinson- Do you happen to have any example pages where this is used in Chrome? I'm wondering if there are any standards to guide the layout of the dialog. 

Of not, then please just enlarge the dialog enough to allow for a larger input field, input text size (min 13px) and some margin around the text.


thanks!

File: app/browser/tabMessageBox.js [hunk 1]
```diff
@@ -45,10 +45,20 @@ const tabMessageBox = {
...
-          shouldDisplaySuppressCheckbox, isBeforeUnloadDialog, isReload, muonCb) => {
-      console.warn('window.prompt is not supported yet')
-      let suppress = false
-      muonCb(null, '', suppress)
+         shouldDisplaySuppressCheckbox, isBeforeUnloadDialog, isReload, muonCb) => {
+      const tabId = webContents.getId()
+      const detail = {
+        message,
+        title,
+        buttons: ['ok', 'cancel'],
```
Comment (NejcZdovc) [inline app/browser/tabMessageBox.js]: Don't forget to add translation for this two buttons

Comment (AlexRobinson-) [reply app/browser/tabMessageBox.js]: Can do! Did you want me to replace the other locations this text appears as well?

E.g.
https://github.com/AlexRobinson-/browser-laptop/blob/ef1cb6caf42f1a79c71715e20bb2f450e58d74b6/app/browser/tabMessageBox.js#L24

https://github.com/AlexRobinson-/browser-laptop/blob/ef1cb6caf42f1a79c71715e20bb2f450e58d74b6/app/browser/tabMessageBox.js#L38

**Comment by AlexRobinson-** [comment]
@bradleyrichter An example page I have been using that was mentioned in the original issue #94 is https://www.w3schools.com/js/tryit.asp?filename=tryjs_prompt

Chrome has the prompt as a full width input which is what I am planning to do. However I'll also make sure to increase the size of the text field as you said

File: app/browser/tabMessageBox.js (hunk 1 above)
Comment (bsclifton) [reply app/browser/tabMessageBox.js]: @AlexRobinson- that would be great 😄 👍

**Comment by AlexRobinson-** [comment]
@bradleyrichter What are your thoughts on this?

image

**Comment by bradleyrichter** [comment]
Perfection!

Cc @bclifton

**Comment by bsclifton** [comment]
@AlexRobinson- anything else required (is this ready for review)? 😄 Wanted to check as the initial checkboxes in top post haven't been updated

**Comment by AlexRobinson-** [comment]
Hey @bsclifton

Only thing left to do is add some tests :)

I struggled with them a bit on my first attempt and unfortunately I have been a bit busy lately, so haven't quite had time to finish these up.

Hopefully I'll have some time this weekend to have another go, but I may need some help with the tests.

Am on mobile so having a hard time linking files, but I tried to copy the tests for other functions (window.confirm) that use HTML files to load up and test the whole flow.

Any advice/help with writing the window.prompt tests are appreciated!

**Comment by immanuelfodor** [comment]
Please somebody help @AlexRobinson- to write those tests to finally merge these changes, the missing `window.prompt` implementation is making me switch back to Chrome regularly, we really need this functionality in Brave :)

**Comment by AlexRobinson-** [comment]
Sorry everyone to hold you all up on this, haven't quite had much time to come back to this.

I have added some unit tests to the `MessageBox` component to reflect its changes

For the actual flow though I have been trying to copy the tests for the confirm dialog here

https://github.com/brave/browser-laptop/blob/master/test/app/renderer/components/messageBoxTest.js#L249

But your way @NejcZdovc sounds a bit easier

**Comment by AlexRobinson-** [comment]
@bsclifton do you think this is ok to be merged in now?

**Comment by bsclifton** [comment]
@AlexRobinson- checking it out now... 😄

**Comment by AlexRobinson-** [comment]
> The only thing missing is that the textbox shown in the prompt should be highlighted when it's shown

Completely fair point! I'll implement this tomorrow night

**Comment by AlexRobinson-** [comment]
@bsclifton All done! I have updated the PR description with a new gif showing this functionality

**Comment by bsclifton** [comment]
@cezaraugusto @petemill would either of you like to give this a review before I merge?

**Comment by bsclifton** [comment]
Uplifted to 0.23.x with 88befe4b52a88581b0fd48dcdf77f73603382485

**Pull request merged**